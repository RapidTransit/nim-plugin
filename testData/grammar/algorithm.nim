proc reverse[T](a: var openArray[T], first, last: Natural) =
  ## reverses the array ``a[first..last]``.
  var x = first
  var y = last
  while x < y:
    swap(a[x], a[y])
    dec(y)
    inc(x)
