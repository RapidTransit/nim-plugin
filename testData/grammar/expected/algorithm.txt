Nim file
  NimProcedureImpl(PROCEDURE)
    PsiElement(NimTokenType.proc)('proc')
    PsiWhiteSpace(' ')
    PsiElement(NimTokenType.IDENTIFIER)('reverse')
    NimGenericTypeImpl(GENERIC_TYPE)
      PsiElement(NimTokenType.[)('[')
      PsiElement(NimTokenType.IDENTIFIER)('T')
      PsiElement(NimTokenType.])(']')
    PsiElement(NimTokenType.()('(')
    NimArgumentsImpl(ARGUMENTS)
      NimArgumentImpl(ARGUMENT)
        PsiElement(NimTokenType.IDENTIFIER)('a')
        PsiElement(NimTokenType.SINGLE_COLON)(':')
        PsiWhiteSpace(' ')
        PsiElement(NimTokenType.var)('var')
        PsiWhiteSpace(' ')
        NimQualifiedTypeImpl(QUALIFIED_TYPE)
          PsiElement(NimTokenType.IDENTIFIER)('openArray')
          NimGenericTypeImpl(GENERIC_TYPE)
            PsiElement(NimTokenType.[)('[')
            PsiElement(NimTokenType.IDENTIFIER)('T')
            PsiElement(NimTokenType.])(']')
      PsiElement(NimTokenType.,)(',')
      PsiWhiteSpace(' ')
      NimArgumentImpl(ARGUMENT)
        PsiElement(NimTokenType.IDENTIFIER)('first')
        PsiElement(NimTokenType.,)(',')
        PsiWhiteSpace(' ')
        PsiElement(NimTokenType.IDENTIFIER)('last')
        PsiElement(NimTokenType.SINGLE_COLON)(':')
        PsiWhiteSpace(' ')
        NimQualifiedTypeImpl(QUALIFIED_TYPE)
          PsiElement(NimTokenType.IDENTIFIER)('Natural')
    PsiElement(NimTokenType.))(')')
    PsiWhiteSpace(' ')
    PsiElement(NimTokenType.=)('=')
  PsiElement(NimTokenType.<crlf>)('\n')
  PsiWhiteSpace(' ')
  PsiWhiteSpace(' ')
  PsiComment(NimTokenType.HASH)('## reverses the array ``a[first..last]``.')
  PsiElement(NimTokenType.<crlf>)('\n')
  PsiWhiteSpace(' ')
  PsiWhiteSpace(' ')
  NimAssignmentImpl(ASSIGNMENT)
    PsiElement(NimTokenType.var)('var')
    PsiWhiteSpace(' ')
    PsiElement(NimTokenType.IDENTIFIER)('x')
    PsiWhiteSpace(' ')
    PsiElement(NimTokenType.=)('=')
    PsiWhiteSpace(' ')
    NimIdentifierOrLiteralImpl(IDENTIFIER_OR_LITERAL)
      PsiElement(NimTokenType.IDENTIFIER)('first')
  PsiElement(NimTokenType.<crlf>)('\n')
  PsiWhiteSpace(' ')
  PsiWhiteSpace(' ')
  NimAssignmentImpl(ASSIGNMENT)
    PsiElement(NimTokenType.var)('var')
    PsiWhiteSpace(' ')
    PsiElement(NimTokenType.IDENTIFIER)('y')
    PsiWhiteSpace(' ')
    PsiElement(NimTokenType.=)('=')
    PsiWhiteSpace(' ')
    NimIdentifierOrLiteralImpl(IDENTIFIER_OR_LITERAL)
      PsiElement(NimTokenType.IDENTIFIER)('last')
  PsiElement(NimTokenType.<crlf>)('\n')
  PsiWhiteSpace(' ')
  PsiWhiteSpace(' ')
  NimComparisonStatementImpl(COMPARISON_STATEMENT)
    PsiElement(NimTokenType.while)('while')
    PsiWhiteSpace(' ')
    NimComparisonConditionImpl(COMPARISON_CONDITION)
      PsiElement(NimTokenType.IDENTIFIER)('x')
      PsiWhiteSpace(' ')
      PsiElement(NimTokenType.COMPARISON)('<')
      PsiWhiteSpace(' ')
      PsiElement(NimTokenType.IDENTIFIER)('y')
    PsiElement(NimTokenType.SINGLE_COLON)(':')
  PsiElement(NimTokenType.<crlf>)('\n')
  PsiWhiteSpace(' ')
  PsiWhiteSpace(' ')
  PsiWhiteSpace(' ')
  PsiWhiteSpace(' ')
  NimMethodCallImpl(METHOD_CALL)
    PsiElement(NimTokenType.IDENTIFIER)('swap')
    PsiElement(NimTokenType.()('(')
    NimIdentifierOrLiteralImpl(IDENTIFIER_OR_LITERAL)
      NimArrayAccessImpl(ARRAY_ACCESS)
        PsiElement(NimTokenType.IDENTIFIER)('a')
        PsiElement(NimTokenType.[)('[')
        NimIdentifierOrLiteralImpl(IDENTIFIER_OR_LITERAL)
          PsiElement(NimTokenType.IDENTIFIER)('x')
        PsiElement(NimTokenType.])(']')
    PsiElement(NimTokenType.,)(',')
    PsiWhiteSpace(' ')
    NimIdentifierOrLiteralImpl(IDENTIFIER_OR_LITERAL)
      NimArrayAccessImpl(ARRAY_ACCESS)
        PsiElement(NimTokenType.IDENTIFIER)('a')
        PsiElement(NimTokenType.[)('[')
        NimIdentifierOrLiteralImpl(IDENTIFIER_OR_LITERAL)
          PsiElement(NimTokenType.IDENTIFIER)('y')
        PsiElement(NimTokenType.])(']')
    PsiElement(NimTokenType.))(')')
  PsiElement(NimTokenType.<crlf>)('\n')
  PsiWhiteSpace(' ')
  PsiWhiteSpace(' ')
  PsiWhiteSpace(' ')
  PsiWhiteSpace(' ')
  NimMethodCallImpl(METHOD_CALL)
    PsiElement(NimTokenType.IDENTIFIER)('dec')
    PsiElement(NimTokenType.()('(')
    NimIdentifierOrLiteralImpl(IDENTIFIER_OR_LITERAL)
      PsiElement(NimTokenType.IDENTIFIER)('y')
    PsiElement(NimTokenType.))(')')
  PsiElement(NimTokenType.<crlf>)('\n')
  PsiWhiteSpace(' ')
  PsiWhiteSpace(' ')
  PsiWhiteSpace(' ')
  PsiWhiteSpace(' ')
  NimMethodCallImpl(METHOD_CALL)
    PsiElement(NimTokenType.IDENTIFIER)('inc')
    PsiElement(NimTokenType.()('(')
    NimIdentifierOrLiteralImpl(IDENTIFIER_OR_LITERAL)
      PsiElement(NimTokenType.IDENTIFIER)('x')
    PsiElement(NimTokenType.))(')')
  PsiErrorElement:Empty element parsed in 'File_0' at offset 193
    <empty list>